OEL-1614: Permissions and roles for pages and contact forms

donquixote opened the pull request.

donquixote committed:
OEL-1614: Allow editor to delete any news.

donquixote committed:
OEL-1614: Allow editor to manage oe_showcase_page content.

donquixote committed:
OEL-1614: Create role 'manage_contact_forms', and enable for roleassign.

donquixote committed:
OEL-1614: Test access to manage contact admin paths.

donquixote committed:
OEL-1614: Allow anonymous and authenticated to submit contact forms.

Update test, now we no longer need to apply custom perms.

donquixote committed:
OEL-1614: Order roles alphabetically.

donquixote committed:
OEL-1614: Enhance doc on UserTrait::createUserWithRoles().

donquixote committed:
OEL-1614: Remove system.site from config_devel entries.

donquixote committed:
OEL-1614: Fix config_devel entries in oe_showcase.info.yml.

donquixote committed:
OEL-1614: Add upgrade path.

donquixote committed:
OEL-1614: Add helpers to test exclusive roles.

drishu commented on tests/src/Traits/PathAccessTrait.php:
```diff
@@ -0,0 +1,85 @@
+<?php
+
+declare(strict_types=1);
```
how did this pass phpcs ?

drishu commented on config/install/user.role.manage_contact_forms.yml:
```diff
@@ -0,0 +1,15 @@
+weight: 4
+is_admin: null
+permissions:
+  - 'administer contact forms'
```
the user has to know the urls by heart, lets add some permissions to view the admin pages, access toolbar

drishu commented on config/install/user.role.manage_contact_forms.yml:
```diff
@@ -0,0 +1,15 @@
+permissions:
+  - 'administer contact forms'
+  - 'export contact form messages'
+  - 'view contact messages'
```
we also agreed that this role (and only this role) would have the "access site-wide contact form", this way after creating the form he/she can already view the form and embed once good and ready

donquixote committed:
OEL-1614: Grant access to admin UI for 'manage_contact_forms'.

drishu reviewed (changes requested):
round 1

donquixote committed:
OEL-1614: Fix namespace in contact form test.

donquixote committed:
OEL-1614: Grant 'access site-wide contact form' to 'manage_contact_forms' role.

This will allow this role to preview contact forms under their canonical url.
Other users will only see that contact form as a paragraph.

donquixote committed:
OEL-1614: Test access to manage content pages.

donquixote committed:
OEL-1614: Update test - editor role is sufficient to create page with contact form paragraph.

donquixote committed:
OEL-1614: Don't reorder roles. Re-save roleassign settings.

brummbar commented on tests/src/Traits/PathAccessTrait.php:
```diff
@@ -0,0 +1,84 @@
+/**
+ * Provides methods to check access to paths.
+ */
+trait PathAccessTrait {
```
This trait contains path assertions, let's rename it accordingly:

```suggestion
trait AssertPathAccessTrait {
```

brummbar commented on tests/src/Traits/PathAccessTrait.php:
```diff
@@ -0,0 +1,84 @@
+  /**
+   * Asserts that only users with given role can visit given paths.
+   *
+   * Side effects:
+   * - Creates user accounts.
+   * - Changes login state.
+   * - Changes currently visited page.
```
This can be seen in the code easily. Let's drop all of these entries, also in the other methods.

donquixote replied:
A method name + doc comment should reflect how the method behaves to the outside.
If the name starts with `assert`, I would typically expect a pure method with no side effect.
This is not the case here, the behavior defies expectation for an assert method - this is worthwhile to mention in a  comment.
If you can propose a different name, I am ready to drop the comment :)

brummbar replied:
Uhm let's keep it indeed.

brummbar commented on tests/src/Traits/PathAccessTrait.php:
```diff
@@ -0,0 +1,84 @@
+    $this->drupalLogin($this->createUserWithRoles($other_roles));
+    $this->assertPathsResponseCode(403, $paths);
+
+    // Test editor.
```
```suggestion
    // Test the specified role.
```

brummbar replied:
Just as a note, it might be that in the future multiple roles can access a route, so we will have to change it.
Since we don't have such scenario now, let's leave this like it is now.

donquixote replied:
+1 for the comment fix.

> Just as a note, it might be that in the future multiple roles can access a route, so we will have to change it.
> Since we don't have such scenario now, let's leave this like it is now.

I was thinking the same.
We should change it based on an actual scenario, otherwise we risk to add a useless dimension of flexibility.

brummbar reviewed (changes requested):
I like the tests!

donquixote committed:
OEL-1614: Rename trait.

donquixote committed:
OEL-1614: Fix comment.
